src/internal/connector/exchange/service_restore.go

```diff
@@ -259,49 +35,82 @@ func RestoreCollections(
 	deets *details.Builder,
 	errs *fault.Bus,
 ) (*support.ConnectorOperationStatus, error) {
+	if len(dcs) == 0 {
+		return support.CreateStatus(ctx, support.Restore, 0, support.CollectionMetrics{}, ""), nil
+	}
+
 	var (
-		directoryCaches = make(map[string]map[path.CategoryType]graph.ContainerResolver)
-		metrics         support.CollectionMetrics
-		userID          string
+		userID         = dcs[0].FullPath().ResourceOwner()
+		directoryCache = make(map[path.CategoryType]graph.ContainerResolver)
+		handlers       = restoreHandlers(ac)
+		metrics        support.CollectionMetrics
 		// TODO policy to be updated from external source after completion of refactoring
 		policy = control.Copy
 		el     = errs.Local()
 	)
 
-	if len(dcs) > 0 {
-		userID = dcs[0].FullPath().ResourceOwner()
-		ctx = clues.Add(ctx, "resource_owner", clues.Hide(userID))
-	}
+	ctx = clues.Add(ctx, "resource_owner", clues.Hide(userID))
 
 	for _, dc := range dcs {
 		if el.Failure() != nil {
 			break
 		}
 
-		userCaches := directoryCaches[userID]
-		if userCaches == nil {
-			directoryCaches[userID] = make(map[path.CategoryType]graph.ContainerResolver)
-			userCaches = directoryCaches[userID]
+		var (
+			isNewCache bool
+			category   = dc.FullPath().Category()
+			ictx       = clues.Add(
+				ctx,
+				"restore_category", category,
+				"restore_full_path", dc.FullPath())
+		)
+
+		handler, ok := handlers[category]
+		if !ok {
+			el.AddRecoverable(clues.New("unsupported restore path category").WithClues(ictx))
+			continue
 		}
 
-		containerID, err := CreateContainerDestination(
-			ctx,
-			creds,
-			dc.FullPath(),
-			dest.ContainerName,
-			userCaches,
+		if directoryCache[category] == nil {
+			directoryCache[category] = handler.newContainerCache(userID)
+			isNewCache = true
+		}
+
+		containerID, gcr, err := createDestination(
+			ictx,
+			handler,
+			handler.formatRestoreDestination(dest.ContainerName, dc.FullPath()),
+			userID,
+			directoryCache[category],
+			isNewCache,
 			errs)
 		if err != nil {
-			el.AddRecoverable(clues.Wrap(err, "creating destination").WithClues(ctx))
+			el.AddRecoverable(err)
 			continue
 		}
 
-		temp, canceled := restoreCollection(ctx, ac, gs, dc, containerID, policy, deets, errs)
+		directoryCache[category] = gcr
+
+		ictx = clues.Add(ictx, "restore_destination_id", containerID)
+
+		temp, err := restoreCollection(
+			ictx,
+			handler,
+			dc,
+			userID,
+			containerID,
+			policy,
+			deets,
+			errs)
 
 		metrics = support.CombineMetrics(metrics, temp)
 
-		if canceled {
-			break
+		if err != nil {
+			if graph.IsErrTimeout(err) {
```

Comment:
I think now it won't stop if it's canceled though :/

Reply:
rofl!  Right, right.  Okay, canceled has been added to the graph timeout check.